src/pruna/config/pre_smash_routines.py

```
@@ -72,113 +75,184 @@ def ensure_device_consistency(model, smash_config):
-    """
-    # algorithm groups are subject to change, make sure we have the latest version
-    from pruna.config.smash_space import ALGORITHM_GROUPS
-
-    # iterate through compiler, quantizer, ...
-    for current_group in ALGORITHM_GROUPS:
-        algorithm = smash_config[current_group]
-        if algorithm is not None:
-            check_algorithm_availability(algorithm, current_group, algorithm_dict)
-            algorithm_dict[current_group][algorithm].pre_smash_hook(model, smash_config)
+    algorithm_order : list[str]
+        The order of the active algorithms to be applied.
+    """
+    active_algorithms = algorithm_order
+    for algorithm in active_algorithms:
+        algorithm_class = AlgorithmRegistry[algorithm]
+        algorithm_class.pre_smash_hook(model, smash_config)
+
+
+def check_algorithm_cross_compatibility(smash_config: SmashConfig) -> None:
+    """
+    Check if the active algorithms are compatible with each other.
+
+    Parameters
+    ----------
+    smash_config : SmashConfig
+        The SmashConfig object containing the algorithm configuration.
+    """
+    active_algorithms = smash_config.get_active_algorithms()
+    algorithm_pairs = list(itertools.permutations(active_algorithms, 2))
+
+    for alg_a, alg_b in algorithm_pairs:
+        if alg_a in AlgorithmRegistry[alg_b].get_incompatible_algorithms():
+            raise ValueError(f"Algorithm {alg_a} is incompatible with {alg_b}")
+
+
+def determine_algorithm_order(smash_config: SmashConfig) -> list[str]:
+    """
+    Determine the order of the active algorithms based on their ordering requirements.
+
+    Parameters
+    ----------
+    smash_config : SmashConfig
+        The SmashConfig object containing the algorithm configuration.
+
+    Returns
+    -------
+    list[str]
+        The order of the active algorithms to be applied.
+    """
+    if smash_config._algorithm_order is not None:
+        return smash_config._algorithm_order
```

I could see an order compatibility check and involving the experimental flag here. What I mean is the difference between these two situation:
- algorithms A and B are compatibile in any order and the user wants to apply them in a specific order: great!
- algorithms A and B are not explicitely compatible: should we raise an error unless the experimental flag is used or should we assume that manually setting _algorithm_order is enough? Should we at least raise a warning?